### ISSUE TITLE
[no-unused-vars] Not working for constructor parameters

### ISSUE THREAD

**Issue by sam-s4s** [typescript-eslint/typescript-eslint#3499]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": "error"
  }
}
```

```TS
private varA: number; // in testing I noticed this isn't being picked up either

constructor(
    varB: number, // this does not get picked up
    private varC: number, // this does not get picked up
    public varD: number // this is fine, and should not be picked up (correct)
) {
    // doesn't use any of the params
}
```

**Expected Result**
Property 'varA' is declared but its value is never read
Parameter 'varB' is declared but its value is never read
Parameter 'varC' is declared but its value is never read (note: not sure if this error should read parameter or property, as it is both)

**Actual Result**
All files pass linting.

**Additional Info**
Many lint errors in my project were picked up using the `no-unused-vars` rule, so I know the rule is active, and working in most cases. But for some reason, none of the above were picked up. Seems strange, as they appear to be pretty straight-forward unused var scenarios.

I notice there is also a rule `no-unused-vars-experimental` that does pick these errors up. Should I just disable the `no-unused-vars` rule and use this instead? Why does the `no-unused-vars` rule exist at all if it is incomplete? Or is the above actually a bug?

Also the existence of the 2 rules is a bit confusing. I'm unsure which of the 2 I should be using - possibly both? Or does the experimental rule cover off all of the cases that the normal rule does, and more, and therefore it's safe to just use that?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.23.0` |
| `@typescript-eslint/parser`        | `4.23.0` |
| `TypeScript`                       | `4.2.4` |
| `ESLint`                           | `7.27.0` |
| `node`                             | `14.17.0` |

**Comment by bradzacher** [comment]
Did you read the rule docs?

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unused-vars.md

They explain what you should be using and link out to the base rule's docs which explain why `varB` isn't caught. Hint: the default of the `args` option.

As for `varA` and `varC` - these are class properties - they are not variables.
They are out of scope of the rule.

**Comment by sam-s4s** [comment]
Thank you for pointing out the up the `'args': 'all'` option, not sure how I missed that, my bad.

Hmmm, I am still left in an uncertain position.

* I am unsure why more people aren't asking about this? Who wouldn't want their unused (private) class properties to be linted, just like other vars? Is it covered by something else that I'm missing?
* If class properties are out of scope of this rule, it seems unintuitive that the experimental version does cover them. It states that the 2 rules are vastly different. Perhaps the name 'no-unused-vars-experimental` is not appropriate, given they are so different?
* I cannot find any other rule that does cover them (unless it has a strange name?).
* The documentation for the experimental rule does not really cover the overlap - for example, I have no way of knowing which of the 2 rules I should be using, or both. (It might just be a documentation issue?)

**Comment by bradzacher** [comment]
> Who wouldn't want their unused (private) class properties to be linted, just like other vars?

Classes in JS aren't used as much as you might expect them to be.
They're up and down - and with the advent of functional react the usage is trending down (from what I know).

> Is it covered by something else that I'm missing?

It is not - nobody has ever asked for it 🤷‍♂️ 

> no-unused-vars-experimental

Back when I wrote it I was experimenting with ways to avoid writing a complete TS scope analyser.
So I built it on top of TS's inbuilt unused vars checker.. as in the rule literally just gets the diagnostics reported by the `noUnusedParams` and `noUnusedLocals` compiler options.

TS covers class properties - so the experimental rule does as well.
But getting TS to calculate diagnostics for a file takes 5-50ms per file.
Which is ***awfully*** slow for a lint run.

Hence I spent the time writing a complete scope analyser in JS to properly support the official `no-unused-vars` rule.

`no-unused-vars-experimental`, as the name suggests, was an experiment that shall be deleted in the next major version.
It is marked as deprecated... but ESLint doesn't do a good job of exposing that to users.

**Comment by sam-s4s** [comment]
Firstly, thank you very much for taking the time to explain :D

Oh, so it sounds like using experimental is a bad idea then.

Is it possible that either class properties could be added to the existing `no-unused-vars` rule, or a new rule created for them? (I could create a new, more focussed issue if that would help).

I'm using Angular, which is completely class based, and this seems like a very handy thing to have. In our projects (several quite large ones), the experimental version finds about 100 instances of unused properties, hah.

**Comment by bradzacher** [comment]
> Is it possible that either class properties could be added to the existing no-unused-vars rule

No. Class properties are not variables any more than object properties are.
So it does not belong in the no-unused-vars rule.

> or a new rule created for them? (I could create a new, more focussed issue if that would help).

There's no reason it couldn't be a new rule. Happy if you raise a new issue!

Fair warning though - it will not an easy thing to build (see below).
As a volunteer maintainer I don't have the spare time, and given you're the first person to ask for this - it's not highly likely another contributor would be motivated to pick it up.

Variables are concrete things - they exist in the AST and can never be referenced ever via a computed name[1].
So `no-unused-vars` is "easy" because you can calculate it purely based on the AST [once you write some code to understand JS variable scoping](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/scope-manager).

The problem you'll find with class properties though is that checking for usage is an impossible problem to solve via the AST - you HAVE to use type information.
This is because `const prop = 'propName'; this[prop];` is valid code to reference a property.
Also it's hard to catch `const self = this; self.propName` via the AST (not impossible, just **hard**).

TS uses its own type info to calculate unused variables.
In a nutshell I believe it's something along the lines of "check all typed member access and explicitly mark which property is being accessed.
Easy for them to build as part of a complete typecheck cycle, but not easy (or fast) to build as part of a lint cycle.

It's also worth noting that there's a harder problem solve in the form of encapsulation breaking member access on class instances:
```ts
class Foo { private prop = 1 }
(new Foo())['prop']; // this is valid TS code for accessing private members!
```
This code is allowed across module boundaries - which makes it very hard to check from a lint run.

TS's logic handles this more easily because it checks every file as part of its typecheck run.

Worth noting that **private names** (`this.#prop`) are an ENTIRELY different story - they are *much* easier to check because computed string access is semantically invalid and so is encapsulation-breaking instance access.
So you can solve this case entirely via the AST! (The only tricky case is `const self = this; self.#prop`)

Not trying to discourage you or anything 😅
I'm just giving you the full picture of the problem for full context.
I think there is value in such a rule - but it will not be easy.

---

[1] not strictly true for `var`s in the global namespace.. but nobody respects that dumb usecase
